File: backend/src/controllers/scadenzeController.ts
import { Request, Response } from 'express';
import * as scadenzeService from '../services/scadenzeService';
import * as pdfService from '../services/pdfService';
import { AppError } from '../middleware/errorHandler';

export async function list(req: Request, res: Response) {
  const filters: {
    cliente_id?: number;
    utente_id?: number;
    filter?: string;
    date_start?: string;
    date_end?: string;
  } = {};

  if (req.query.cliente_id) {
    const parsed = Number(req.query.cliente_id);
    if (isNaN(parsed)) throw new AppError(400, 'cliente_id must be a number');
    filters.cliente_id = parsed;
  }

  if (req.query.utente_id) {
    const parsed = Number(req.query.utente_id);
    if (isNaN(parsed)) throw new AppError(400, 'utente_id must be a number');
    filters.utente_id = parsed;
  }

  if (req.query.filter) {
    filters.filter = String(req.query.filter);
  }

  if (req.query.date_start) {
    const ds = String(req.query.date_start);
    if (isNaN(Date.parse(ds))) throw new AppError(400, 'date_start must be a valid date');
    filters.date_start = ds;
  }

  if (req.query.date_end) {
    const de = String(req.query.date_end);
    if (isNaN(Date.parse(de))) throw new AppError(400, 'date_end must be a valid date');
    filters.date_end = de;
  }

  const data = await scadenzeService.getAllScadenze(
    Object.keys(filters).length > 0 ? filters : undefined
  );

  res.json({ data });
}

export async function getById(req: Request, res: Response) {
  const id = Number(req.params.id);
  if (isNaN(id)) throw new AppError(400, 'id must be a number');

  const data = await scadenzeService.getScadenzaById(id);
  res.json({ data });
}

export async function create(req: Request, res: Response) {
  const { name, rec, date, utente_ids, cliente_ids } = req.body;

  if (!name || typeof name !== 'string' || name.trim().length === 0) {
    throw new AppError(400, 'name is required and must be a non-empty string');
  }

  if (rec !== undefined && (typeof rec !== 'number' || rec < 0)) {
    throw new AppError(400, 'rec must be a non-negative number');
  }

  if (date !== undefined && isNaN(Date.parse(date))) {
    throw new AppError(400, 'date must be a valid date string');
  }

  if (utente_ids !== undefined) {
    if (!Array.isArray(utente_ids) || !utente_ids.every((id: any) => typeof id === 'number')) {
      throw new AppError(400, 'utente_ids must be an array of numbers');
    }
  }

  if (cliente_ids !== undefined) {
    if (!Array.isArray(cliente_ids) || !cliente_ids.every((id: any) => typeof id === 'number')) {
      throw new AppError(400, 'cliente_ids must be an array of numbers');
    }
  }

  const data = await scadenzeService.createScadenza(
    {
      name: name.trim(),
      rec,
      date,
      utente_ids,
      cliente_ids,
    },
    req.user?.id
  );

  res.status(201).json({ data });
}

export async function update(req: Request, res: Response) {
  const id = Number(req.params.id);
  if (isNaN(id)) throw new AppError(400, 'id must be a number');

  const { name, rec, date, utente_ids, cliente_ids } = req.body;

  // At least one field must be provided
  if (
    name === undefined &&
    rec === undefined &&
    date === undefined &&
    utente_ids === undefined &&
    cliente_ids === undefined
  ) {
    throw new AppError(400, 'At least one field must be provided for update');
  }

  if (name !== undefined && (typeof name !== 'string' || name.trim().length === 0)) {
    throw new AppError(400, 'name must be a non-empty string');
  }

  if (rec !== undefined && (typeof rec !== 'number' || rec < 0)) {
    throw new AppError(400, 'rec must be a non-negative number');
  }

  if (date !== undefined && date !== null && isNaN(Date.parse(date))) {
    throw new AppError(400, 'date must be a valid date string or null');
  }

  if (utente_ids !== undefined) {
    if (!Array.isArray(utente_ids) || !utente_ids.every((id: any) => typeof id === 'number')) {
      throw new AppError(400, 'utente_ids must be an array of numbers');
    }
  }

  if (cliente_ids !== undefined) {
    if (!Array.isArray(cliente_ids) || !cliente_ids.every((id: any) => typeof id === 'number')) {
      throw new AppError(400, 'cliente_ids must be an array of numbers');
    }
  }

  const data = await scadenzeService.updateScadenza(id, {
    name: name?.trim(),
    rec,
    date,
    utente_ids,
    cliente_ids,
  });

  res.json({ data });
}

export async function remove(req: Request, res: Response) {
  const id = Number(req.params.id);
  if (isNaN(id)) throw new AppError(400, 'id must be a number');

  const hard = req.query.hard === 'true';

  await scadenzeService.deleteScadenza(id, hard);
  res.status(204).send();
}

export async function exportPDF(req: Request, res: Response) {
  const filters: {
    cliente_id?: number;
    utente_id?: number;
    filter?: string;
    date_start?: string;
    date_end?: string;
  } = {};

  if (req.query.cliente_id) {
    const parsed = Number(req.query.cliente_id);
    if (isNaN(parsed)) throw new AppError(400, 'cliente_id must be a number');
    filters.cliente_id = parsed;
  }

  if (req.query.utente_id) {
    const parsed = Number(req.query.utente_id);
    if (isNaN(parsed)) throw new AppError(400, 'utente_id must be a number');
    filters.utente_id = parsed;
  }

  if (req.query.filter) {
    filters.filter = String(req.query.filter);
  }

  if (req.query.date_start) {
    const ds = String(req.query.date_start);
    if (isNaN(Date.parse(ds))) throw new AppError(400, 'date_start must be a valid date');
    filters.date_start = ds;
  }

  if (req.query.date_end) {
    const de = String(req.query.date_end);
    if (isNaN(Date.parse(de))) throw new AppError(400, 'date_end must be a valid date');
    filters.date_end = de;
  }

  const scadenze = await scadenzeService.getAllScadenze(
    Object.keys(filters).length > 0 ? filters : undefined
  );

  const pdfBuffer = await pdfService.generateScadenzePDF(scadenze);

  // Set PDF response headers
  const today = new Date().toISOString().split('T')[0];
  res.setHeader('Content-Type', 'application/pdf');
  res.setHeader('Content-Disposition', `attachment; filename="scadenze_${today}.pdf"`);
  res.setHeader('Content-Length', pdfBuffer.length);

  res.send(pdfBuffer);
}

File: backend/Controllers/ScadenzeController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Backend.Data;
using Backend.Models;
using Backend.Services;
using System.Globalization;

namespace Backend.Controllers;

[ApiController]
[Route("api/v1/scadenze")]
public class ScadenzeController : ControllerBase
{
    private readonly AppDbContext _context;

    public ScadenzeController(AppDbContext context)
    {
        _context = context;
    }

    [HttpGet]
    public async Task<IActionResult> GetScadenze(
        [FromQuery] int? cliente_id,
        [FromQuery] int? utente_id,
        [FromQuery] string? filter,
        [FromQuery] string? date_start,
        [FromQuery] string? date_end)
    {
        var query = _context.Scadenze.AsQueryable();

        if (cliente_id.HasValue)
        {
            query = query.Where(s => s.ScadenzeClienti.Any(sc => sc.ClienteId == cliente_id.Value));
        }

        if (utente_id.HasValue)
        {
            query = query.Where(s => s.ScadenzeUtenti.Any(su => su.UtenteId == utente_id.Value));
        }

        if (!string.IsNullOrEmpty(filter))
        {
            query = query.Where(s => s.Name.Contains(filter));
        }

        // Parse date filters for calculated date filtering
        DateTime? filterDateStart = null;
        DateTime? filterDateEnd = null;
        if (!string.IsNullOrEmpty(date_start) && DateTime.TryParse(date_start, out var startDate))
        {
            filterDateStart = startDate.Date;
        }
        if (!string.IsNullOrEmpty(date_end) && DateTime.TryParse(date_end, out var endDate))
        {
            filterDateEnd = endDate.Date;
        }

        var scadenzeRaw = await query
            .Select(s => new {
                s.Id,
                s.Name,
                s.Date,
                s.Rec,
                s.InsDate,
                s.ModDate,
                s.InsUserId,
                s.IsActive,
                ScadenzeUtenti = s.ScadenzeUtenti.Select(su => new { su.Id, su.UtenteId, su.ScadenzaId, su.Utente.Nome, su.Utente.Cognome }),
                ScadenzeClienti = s.ScadenzeClienti.Select(sc => new { sc.Id, sc.ClienteId, sc.ScadenzaId, sc.Cliente.Name })
            })
            .ToListAsync();

        // Calculate real date and filter by date range
        var today = DateTime.UtcNow.Date;
        var scadenze = scadenzeRaw
            .Select(s => {
                DateTime? scadenzaReale = s.Date;
                if (s.Date.HasValue && s.Rec > 0)
                {
                    var calcDate = s.Date.Value;
                    while (calcDate.Date < today)
                    {
                        calcDate = calcDate.AddMonths(s.Rec);
                    }
                    scadenzaReale = calcDate;
                }
                return new {
                    s.Id,
                    s.Name,
                    s.Date,
                    ScadenzaReale = scadenzaReale,
                    s.Rec,
                    s.InsDate,
                    s.ModDate,
                    s.InsUserId,
                    s.IsActive,
                    s.ScadenzeUtenti,
                    s.ScadenzeClienti
                };
            })
            .Where(s => {
                if (!s.ScadenzaReale.HasValue) return false;
                var realDate = s.ScadenzaReale.Value.Date;
                if (filterDateStart.HasValue && realDate < filterDateStart.Value) return false;
                if (filterDateEnd.HasValue && realDate > filterDateEnd.Value) return false;
                return true;
            })
            .OrderBy(s => s.ScadenzaReale)
            .ToList();

        return Ok(scadenze);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetScadenza(int id)
    {
        var scadenza = await _context.Scadenze.FindAsync(id);
        if (scadenza == null) return NotFound();
        return Ok(scadenza);
    }

    [HttpPost]
    public async Task<IActionResult> CreateScadenza(Scadenze scadenza)
    {
        _context.Scadenze.Add(scadenza);
        await _context.SaveChangesAsync();
        return CreatedAtAction(nameof(GetScadenza), new { id = scadenza.Id }, scadenza);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateScadenza(int id, Scadenze scadenza)
    {
        if (id != scadenza.Id) return BadRequest();
        _context.Entry(scadenza).State = EntityState.Modified;
        await _context.SaveChangesAsync();
        return NoContent();
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteScadenza(int id)
    {
        var scadenza = await _context.Scadenze.FindAsync(id);
        if (scadenza == null) return NotFound();
        _context.Scadenze.Remove(scadenza);
        await _context.SaveChangesAsync();
        return NoContent();
    }

    [HttpGet("export/pdf")]
    public async Task<IActionResult> ExportPdf(
        [FromQuery] int? cliente_id,
        [FromQuery] int? utente_id,
        [FromQuery] string? filter,
        [FromQuery] string? date_start,
        [FromQuery] string? date_end)
    {
        // Get all scadenze with related data
        var query = _context.Scadenze
            .Include(s => s.ScadenzeUtenti).ThenInclude(su => su.Utente)
            .Include(s => s.ScadenzeClienti).ThenInclude(sc => sc.Cliente)
            .AsQueryable();

        // Apply cliente filter
        if (cliente_id.HasValue)
        {
            query = query.Where(s => s.ScadenzeClienti.Any(sc => sc.ClienteId == cliente_id.Value));
        }

        // Apply utente filter
        if (utente_id.HasValue)
        {
            query = query.Where(s => s.ScadenzeUtenti.Any(su => su.UtenteId == utente_id.Value));
        }

        // Apply name filter
        if (!string.IsNullOrEmpty(filter))
        {
            query = query.Where(s => s.Name.Contains(filter));
        }

        var scadenze = await query.ToListAsync();

        // Parse date filters
        DateTime? filterDateStart = null;
        DateTime? filterDateEnd = null;

        if (!string.IsNullOrEmpty(date_start) && DateTime.TryParse(date_start, out var parsedStart))
        {
            filterDateStart = parsedStart;
        }
        if (!string.IsNullOrEmpty(date_end) && DateTime.TryParse(date_end, out var parsedEnd))
        {
            filterDateEnd = parsedEnd;
        }

        // Calculate RO_SCADENZA_REALE (real due date with recurrence) and filter
        var pdfItems = new List<ScadenzaPdfItem>();
        var today = DateTime.UtcNow.Date;

        foreach (var scadenza in scadenze)
        {
            if (!scadenza.Date.HasValue) continue;

            // Calculate real due date with recurrence
            var scadenzaReale = scadenza.Date.Value;

            if (scadenza.Rec > 0)
            {
                // Keep adding REC months until the date is >= today
                while (scadenzaReale.Date < today)
                {
                    scadenzaReale = scadenzaReale.AddMonths(scadenza.Rec);
                }
            }

            // Filter by date range on the calculated real date
            if (filterDateStart.HasValue && scadenzaReale.Date < filterDateStart.Value.Date)
            {
                continue;
            }
            if (filterDateEnd.HasValue && scadenzaReale.Date > filterDateEnd.Value.Date)
            {
                continue;
            }

            // Build clienti string (comma-separated names)
            var clientiNames = scadenza.ScadenzeClienti
                .Where(sc => sc.Cliente != null)
                .Select(sc => sc.Cliente!.Name)
                .Distinct()
                .ToList();

            // Build utenti string (comma-separated cognome)
            var utentiNames = scadenza.ScadenzeUtenti
                .Where(su => su.Utente != null)
                .Select(su => su.Utente!.Cognome)
                .Distinct()
                .ToList();

            pdfItems.Add(new ScadenzaPdfItem
            {
                Name = scadenza.Name,
                Clienti = string.Join(", ", clientiNames),
                Utenti = string.Join(", ", utentiNames),
                ScadenzaReale = scadenzaReale
            });
        }

        // Sort by real due date and assign index
        pdfItems = pdfItems.OrderBy(x => x.ScadenzaReale).ToList();
        for (int i = 0; i < pdfItems.Count; i++)
        {
            pdfItems[i].Index = i + 1;
        }

        // Format date strings for header
        var dateStartStr = filterDateStart?.ToString("dd/MM/yyyy") ?? "--";
        var dateEndStr = filterDateEnd?.ToString("dd/MM/yyyy") ?? "--";

        // Generate PDF
        var pdfService = new PdfExportService();
        var pdfBytes = pdfService.GenerateScadenzePdf(pdfItems, dateStartStr, dateEndStr);

        return File(pdfBytes, "application/pdf", $"scadenze_{DateTime.Now:yyyyMMdd}.pdf");
    }
}